### PR TITLE
Performance improvements to events, stats and service_checks

### PR DESCRIPTION
**Pull request by deepakprabhakara**
This commit reduces allocations across the code by writing directly to the Buffer stream instead of creating intermediate vars. It also safeguards g.Tags from being modified by append calls. This is done by using a temp array sized according to the number of tags set. Also added a benchmark test for service_checks and events.

As an example the events allocation(/op) has dropped from 25 to 3.

New:-
```
13:19:41 DeepakMacbook: > go test -check.b -check.bmem
PASS: events_test.go:189: TestSuite.BenchmarkEvents	  500000	      5566 ns/op	     304 B/op	       3 allocs/op
PASS: stats_test.go:223: TestSuite.BenchmarkIncr	  500000	      5504 ns/op	     336 B/op	       7 allocs/op
OK: 2 passed
PASS
ok  	github.com/redsift/hilt/vendor/github.com/PagerDuty/godspeed	5.788s
```

Old:-
```
13:22:45 DeepakMacbook: > go test -check.b -check.bmem
PASS: events_test.go:189: TestSuite.BenchmarkEvents	  200000	     11714 ns/op	   12962 B/op	      25 allocs/op
PASS: stats_test.go:223: TestSuite.BenchmarkIncr	  200000	      6693 ns/op	     577 B/op	       8 allocs/op
OK: 2 passed
PASS
ok  	github.com/PagerDuty/godspeed	4.009s
```

Note: This could be further optimised by not creating additional intermediate arrays (it can be directly written to stream) and avoiding strings.Join (these can be directly written to stream as well).

### CODE REVIEW

File: service_checks_test.go [hunk 1]
```diff
@@ -114,3 +114,12 @@ func (t *TestSuite) TestServiceCheck(c *C) {
...
 }
+
+func (t *TestSuite) BenchmarkSvcChecks(c *C) {
+	fields := make(map[string]string)
+	fields["service_check_message"] = "server on fire"
+
+	for i := 0; i < c.N; i++ {
+		t.g.ServiceCheck("testSvc", 3, fields, []string{"tag:test", "tag2:testing"})
+	}
+}
```
Comment (theckman) [inline service_checks_test.go]: Are you able to insert a newline here?

**Review by theckman — DISMISSED** [review]
Thank you for this contribution. TravisCI finally recovered enough that CI could run.

I found one minor change, and then I'm happy to merge this PR.

File: service_checks_test.go (hunk 1 above)
Comment (deepakprabhakara) [reply service_checks_test.go]: Sure thing.

Comment (deepakprabhakara) [reply service_checks_test.go]: Shall I just commit this new change and you'll squash the merge?

**Comment by theckman** [comment]
@deepakprabhakara as long as I can do it without affecting your authorship of the commit, that's no problem. I'm pretty sure it'll keep you as the author and me as the committer, which is 👍 by me.

I hope to be back at my workstation in a few hours so that I can give this one final once-over and then merge. Thank you again.

**Review by theckman — APPROVED** [review]

File: events_test.go [hunk 2]
```diff
@@ -8,7 +8,8 @@ import (
 	"fmt"
 	"time"
 
-	// this is *C comes from
+	"github.com/PagerDuty/godspeed"
```
Comment (akrylysov) [inline events_test.go:11]: Wrong import? Should it be `github.com/theckman/godspeed`?

File: events.go [hunk 3]
```diff
@@ -39,26 +38,36 @@ func (g *Godspeed) Event(title, text string, fields map[string]string, tags []st
...
+	buf.WriteString(text)
 
 	// if some fields were passed in convert them to their proper format
 	// and write that to the buffer
 	if len(fields) > 0 {
 		for i, v := range eventKeys {
 			if mv, ok := fields[v]; ok {
-				buf.WriteString(fmt.Sprintf("|%v:%v", string(eventMarkers[i]), removePipes(mv)))
+				buf.WriteString("|")
+				buf.WriteString(string(eventMarkers[i]))
```
Comment (akrylysov) [inline events.go:56]: You can do `buf.WriteRune(eventMarkers[i])` instead.

**Comment by akrylysov** [comment]
Sorry, I'm a little late with my review.

**Comment by theckman** [comment]
@akrylysov Thank you so much for taking time to look at this PR too, and no worries! I found it the hard way (local tests exploded)... 😉 

* https://github.com/theckman/godspeed/pull/5

Thank you again.

@deepakprabhakara I'll be cutting `v1.1.0` soon, which will include these changes. I found some further optimizations I can make (will happen in PR #6 ), and I'll cut `v1.1.0` after I get a chance to merge them in.

File: events_test.go (hunk 2 above)
Comment (deepakprabhakara) [reply events_test.go:11]: Oops, thanks for fixing it @akrylysov @theckman . I blame Goland and goimports! ;-)

**Comment by deepakprabhakara** [comment]
Thanks @theckman @akrylysov for reviewing and taking in the PR. It was mainly based off of @akrylysov work.

**Pull request closed without merging**